eth/db/storage.py

```diff
@@ -0,0 +1,477 @@
+        return hashed_slot in read_trie
+
+    def __delitem__(self, key: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        try:
+            del write_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to be able to delete slot, but:", exc_info=True)
+
+    @property
+    def has_changed_root(self) -> bool:
+        return self._write_trie and self._write_trie.root_hash != self._starting_root_hash
+
+    def get_changed_root(self) -> Hash32:
+        if self._write_trie is not None:
+            return self._write_trie.root_hash
+        else:
+            raise ValidationError("Asked for changed root when no writes have been made")
+
+    def clear_changed_root(self) -> None:
+        self._write_trie = None
+        self._starting_root_hash = None
+
+    def commit_to(self, db: BaseDB) -> None:
+        self.logger.debug2('persist storage root to data store')
+        self._trie_nodes_batch.commit_to(db, apply_deletes=False)
+        self._trie_nodes_batch.clear()
+
+
+class AccountStorageDB:
+    """
+    Storage cache and write batch for a single account. Changes are not
+    merklized until :meth:`make_storage_root` is called.
+    """
+    logger = cast(ExtendedDebugLogger, logging.getLogger("eth.db.storage.AccountStorageDB"))
+
+    def __init__(self, db: BaseAtomicDB, account_db: BaseAccountDB, address: Address) -> None:
+        """
+        Database entries go through several pipes, like so...
+
+        .. code::
+
+            db -> _storage_lookup -> _storage_cache -> _journal_storage
+
+        db is the raw database, we can assume it hits disk when written to.
+        Keys are stored as node hashes and rlp-encoded node values.
+
+        _storage_lookup is itself a pair of databases: (BatchDB -> HexaryTrie),
+        writes to storage lookup *are* immeditaely applied to a trie, generating
+        the appropriate trie nodes and and root hash (via the HexaryTrie). The
+        writes are *not* persisted to db, until _storage_lookup is explicitly instructed to,
+        via :meth:`StorageLookup.commit_to`
+
+        _storage_cache is a cache tied to the state root of the trie. It
+        is important that this cache is checked *after* looking for
+        the key in _journal_storage, because the cache is only invalidated
+        after a state root change. Otherwise, you will see data since the last
+        storage root was calculated.
+        Keys are set/retrieved as the big_endian encoding of the slot integer, and the rlp-encoded
+        value.
+
+        Journaling batches writes at the _journal_storage layer, until persist is called.
+        It manages all the checkpointing and rollbacks that happen during EVM execution.
+        Keys are set/retrieved as the big_endian encoding of the slot integer, and the rlp-encoded
+        value.
```

Comment:
This line: "Keys are set/retrieved as the big_endian encoding of the slot integer, and the rlp-encoded value." appears twice in the above docstring.

Reply:
Hm, one refers to `_storage_cache` and the other refers to `_journal_storage`. Any idea how to make that clearer?

Reply:
Ahh, gotcha, I guess it's okay. Two ideas:
- since `_storage_cache` is never written to by this class and readers should know that all operations to `_journal_storage` will pass through to it, maybe it's fine to drop the first duplication of this line and not specify how keys are encoded.
- Maybe it could become "In both _storage_cache and _journal_storage, keys are set/retreived ..."